### PR TITLE
Vehicle Number Plate handling

### LINKED ISSUE #244: feat: Vehicle Licence Plate Delivery
To ensure that the delivery of the Vehicle license/Registration Plate needs to be properly documented in the ERP, the user should create a delivery confirmation or note from the registration receipt. This will serve as proof of delivery and can be used to track the progress of license/registration plate delivery in the system. It's important to keep this document safe and accessible in case any issues arise during the delivery process or post-delivery.

**Solution**

We need a Licence Plate Delivery document linked with the registration order and receipt with the following details 

- Registration Customer (Fetch from Registration Order)
-  Registration Details (Registration no) - (Fetch from Registration Receipt)
- Vehicle Details (Fetch from Registration Order)
- Receiver Details (by default it'll auto-select the registration customer - but if there is anyone else on the behalf of customer then we need to add the secondary contact person name)
- Submittable document  
-  Printable document

### PR DESCRIPTION
**Pull request by haiderk0987**
Closes https://github.com/ParaLogicTech/erpnext/issues/244

### CODE REVIEW

File: erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.json [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+ "index_web_pages_for_search": 1,
+ "is_submittable": 1,
+ "links": [],
+ "modified": "2023-09-05 18:50:05.286190",
+ "modified_by": "Administrator",
+ "module": "Vehicles",
+ "name": "Vehicle Number Plate Delivery",
+ "naming_rule": "By \"Naming Series\" field",
+ "owner": "Administrator",
+ "permissions": [
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.json:476]: Update permissions ... similar to vehicle invoice delivery

File: erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.json [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+   "search_index": 1
+  },
+  {
+   "fieldname": "serial_no",
+   "fieldtype": "Link",
+   "hidden": 1,
+   "label": "Serial No",
+   "no_copy": 1,
+   "options": "Serial No",
+   "read_only": 1
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.json]: Don't need this field .. we only need this field for stock related documents

File: erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		self.update_vehicle_registration_order()
+
+	def on_cancel(self):
+		self.update_vehicle_registration_order()
+
+	def validate_duplicate_number_plate_delivery(self):
+		filters = {
+			"vehicle": self.vehicle,
+			"docstatus": 1,
+			"vehicle_license_plate": self.vehicle_license_plate,
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py]: No need specify reg # in filter .. One Vehicle one Number Plate delivery would be fine

File: erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+			"name": ['!=', self.name]
+		}
+
+		number_plate_delivery = frappe.db.get_value("Vehicle Number Plate Delivery", filters=filters)
+
+		if number_plate_delivery:
+			frappe.throw(_("Number Plate for {0} has already been delivered in {1}")
+				.format(frappe.get_desk_link("Vehicle", self.vehicle),
+				frappe.get_desk_link("Vehicle Number Plate Delivery", number_plate_delivery)))
+
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py]: indentation is not good

File: erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.json [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+ "index_web_pages_for_search": 1,
+ "is_submittable": 1,
+ "links": [],
+ "modified": "2023-09-06 13:00:27.168409",
+ "modified_by": "Administrator",
+ "module": "Vehicles",
+ "name": "Vehicle Number Plate Receipt",
+ "naming_rule": "By \"Naming Series\" field",
+ "owner": "Administrator",
+ "permissions": [
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.json:97]: Set permissions similar to Vehicle Invoice (Receipt)

File: erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+			if d.vehicle:
+				if d.vehicle in vehicles:
+					frappe.throw(_("Row #{0}: Vehicle {1} is duplicate").format(d.idx, d.vehicle))
+				vehicles.add(d.vehicle)
+
+	def validate_number_plate_already_received(self):
+		for d in self.number_plates:
+			number_plate_received = frappe.db.get_value("Vehicle Number Plate Details", filters={
+				"vehicle": d.vehicle,
+				"vehicle_license_plate": d.vehicle_license_plate,
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.py]: no need to specify reg # .. one vehicle one number plate receipt is fine

File: erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+				"docstatus": 1,
+				"name": ['!=', d.name]
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.py]: `parent != self.name`

File: erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.json [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  "company",
+  "naming_series",
+  "remarks",
+  "column_break_qebot",
+  "posting_date",
+  "agent",
+  "agent_name",
+  "section_break_8kitr",
+  "number_plates",
+  "amended_from"
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_number_plate_receipt/vehicle_number_plate_receipt.json]: Amended from in the first section bottom right

File: erpnext/vehicles/doctype/vehicle_registration_order/vehicle_registration_order.py [hunk 9]
```diff
@@ -573,6 +574,24 @@ def set_registration_receipt_details(self, update=False):
 				"title": self.title,
 			})
 
+	def set_number_plate_status(self, update=False):
```
Comment (SaiFi0102) [inline erpnext/vehicles/doctype/vehicle_registration_order/vehicle_registration_order.py]: argument `update_modified=True`

**Review by SaiFi0102 — CHANGES_REQUESTED** [review]

File: erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py (hunk 3 above)
Comment (haiderk0987) [reply erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py]: Ain't there a possibility for another reg #?

File: erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py (hunk 4 above)
Comment (haiderk0987) [reply erpnext/vehicles/doctype/vehicle_number_plate_delivery/vehicle_number_plate_delivery.py]: ...

**Review by SaiFi0102 — APPROVED** [review]

**Comment by SaiFi0102** [comment]
Changes: https://github.com/ParaLogicTech/erpnext/commit/9e0f00ce46b3cb9683db2ee88887a4fc0cbe7924

**Pull request merged**